server/src/inmantals/server.py

```diff
@@ -229,6 +251,54 @@ async def textDocument_references(self, textDocument, position, context):  # noq
 
         return [self.convert_location(loc.data) for loc in range]
 
+    async def workspace_symbol(self, query: str) -> List[Dict[str, object]]:
+        if self.types is None:
+            return []
+
+        query_upper: str = query.upper()
+        matching_types: List[Tuple[str, inmanta_type.NamedType]] = [
+            (name, tp)
+            for name, tp in self.types.items()
+            if isinstance(tp, inmanta_type.NamedType) and query_upper in name.upper()
+        ]
+
+        def get_symbol_kind(tp: inmanta_type.NamedType) -> lsp_types.SymbolKind:
+            def if_supported(then: lsp_types.SymbolKind, otherwise: lsp_types.SymbolKind) -> lsp_types.SymbolKind:
+                return then if (self.supported_symbol_kinds is not None and then in self.supported_symbol_kinds) else otherwise
+
+            if isinstance(tp, Plugin):
+                return lsp_types.SymbolKind.Function
+            if isinstance(tp, inmanta_type.ConstraintType):
+                return lsp_types.SymbolKind.Class
+            if isinstance(tp, EntityLike):
+                return lsp_types.SymbolKind.Class
+            if isinstance(tp, Implementation):
+                return lsp_types.SymbolKind.Constructor
+
+            logger.exception("Unknown type %s, using default symbol kind" % tp)
+            return if_supported(lsp_types.SymbolKind.Object, lsp_types.SymbolKind.Variable)
```

Comment:
I tried my best to pick the `SymbolKind` that I found best matched the symbol type in the inmanta DSL but most of them are far from a perfect fit. Let me know if you think there are better matches possible.